### PR TITLE
nixos/nats: init

### PR DESCRIPTION
**Pull request by c0deaddict**
###### Motivation for this change

NixOS is missing a module for the NATS message bus service.

###### Things done

Started off with a copy of the mosquitto service module. I don't have much experience with the systemd hardening flags, would like a second opinion on this. The service is running, i've tested it with normal pub/sub, JetStream and MQTT support. Also included a NixOS test that does a basic pub/sub.

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/nats.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/nats.nix:36]: Why do we need this options?

File: nixos/modules/services/networking/nats.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/nats.nix:65]:
```suggestion
        type = types.port;
```

File: nixos/modules/services/networking/nats.nix
Comment (aanderse) [inline nixos/modules/services/networking/nats.nix:109]: I would suggest **not** using `mkDefault` here. All of these values are set by the sysadmins already, so if they want to override they should probably be explicit about it and use `lib.mkForce` in their configuration.

File: nixos/modules/services/networking/nats.nix
Comment (aanderse) [inline nixos/modules/services/networking/nats.nix:94]: :heart: clickable links.

File: nixos/tests/nats.nix
Comment (aanderse) [inline nixos/tests/nats.nix:27]: Oh no! :scream: 

What options does a sysadmin have for not storing passwords in the nix store? How sensitive is this password?

**Review by aanderse — CHANGES_REQUESTED** [review]
Let me know if you have any questions or need a hand with any of this. Full bonus points awarded for the NixOS test!! :tada:

File: nixos/modules/services/networking/nats.nix
Comment (c0deaddict) [reply nixos/modules/services/networking/nats.nix:36]: NATS JetStream can use file storage to store messages in it's distributed streams. I thought a user account is needed for the systemd unit in this case, since it needs a stable uid between restarts. Or can this also be done with `DynamicUser`?

File: nixos/modules/services/networking/nats.nix
Comment (c0deaddict) [reply nixos/modules/services/networking/nats.nix:82]: I wasn't aware of `pkgs.formats.json`, what is the difference with `builtins.toJSON`? Can I use it like in this module: https://github.com/NixOS/nixpkgs/blob/8284fc30c84ea47e63209d1a892aca1dfcd6bdf3/nixos/modules/services/logging/vector.nix#L20 ?

**Comment by c0deaddict** [comment]
> Let me know if you have any questions or need a hand with any of this. Full bonus points awarded for the NixOS test!! tada

Thanks, will do! I really love how powerful and simple to use the NixOS test framework is.

File: nixos/modules/services/networking/nats.nix
Comment (c0deaddict) [reply nixos/modules/services/networking/nats.nix:71]: Yes, correct. Nats core doesn't store any state (that i'm aware of).

File: nixos/modules/services/networking/nats.nix
Comment (c0deaddict) [reply nixos/modules/services/networking/nats.nix:19]: No, there isn't. Personally I always like the availability of a `package` option so i'm able to pin a specific package version. But now that I write this, just realized that this can be done via an overlay as well :sweat_smile:

File: nixos/modules/services/networking/nats.nix
Comment (c0deaddict) [reply nixos/modules/services/networking/nats.nix:167]: Is there a recommended alternative for `createHome = true` but then at service activation? With `createHome = false` the test fails because of: ```server # [    9.079281] systemd[893]: nats.service: Failed to set up mount namespacing: /run/systemd/unit-root/var/lib/nats: No such file or directory```

File: nixos/modules/services/networking/nats.nix
Comment (aanderse) [reply nixos/modules/services/networking/nats.nix:36]: If you use `StateDirectory` in your `systemd` service it will ensure proper ownership every time. The caveat is that your data directory must reside under `/var/lib`.

File: nixos/modules/services/networking/nats.nix
Comment (aanderse) [reply nixos/modules/services/networking/nats.nix:71]: Sounds good. Might be worth mentioning here... unless that is obvious to users of this software.

File: nixos/modules/services/networking/nats.nix
Comment (aanderse) [reply nixos/modules/services/networking/nats.nix:19]: Exactly :+1:

File: nixos/modules/services/networking/nats.nix
Comment (aanderse) [reply nixos/modules/services/networking/nats.nix:82]: [RFC 0042](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md) has been promoting this sort of pattern:

```
  format = pkgs.formats.json {};
  configFile = format.generate "foo.json" cfg.settings;
...
  settings = mkOption {
    type = format.type;
...
  };
```

Under the hood it still uses `toJSON`, so there is no *real* difference in implementation here... just convention and standardization. I think it looks nice.

File: nixos/modules/services/networking/nats.nix
Comment (aanderse) [reply nixos/modules/services/networking/nats.nix:167]: You still need to create the directory. I generally recommend `StateDirectory` to provision these sorts of directories. Check out this:

https://github.com/NixOS/nixpkgs/blob/b72bde7c4a1f9c9bf1a161f0c267186ce3c6483c/nixos/modules/services/misc/gitolite.nix#L26-L35

https://github.com/NixOS/nixpkgs/blob/b72bde7c4a1f9c9bf1a161f0c267186ce3c6483c/nixos/modules/services/misc/gitolite.nix#L165-L177

How does that look to you?

File: nixos/tests/nats.nix
Comment (c0deaddict) [reply nixos/tests/nats.nix:27]: Nats supports [bcrypted passwords](https://docs.nats.io/nats-server/configuration/securing_nats/auth_intro/username_password#bcrypted-passwords) in this field as well as plain text. How sensitive it is depends on the [permissions](https://docs.nats.io/nats-server/configuration/securing_nats/authorization) of the user. Besides basic username/password authentication Nats also supports [client TLS certificates](https://docs.nats.io/nats-server/configuration/securing_nats/auth_intro/tls_mutual_auth) or [NKeys which is based on Ed25519](https://docs.nats.io/nats-server/configuration/securing_nats/auth_intro/nkey_auth). I don't have any experience with these last two options, but it should be configurable with the generic settings JSON.

File: nixos/modules/services/networking/nats.nix
Comment (c0deaddict) [reply nixos/modules/services/networking/nats.nix:82]: Thanks for explaining, it does look nicer to me as well :)

File: nixos/tests/nats.nix
Comment (aanderse) [reply nixos/tests/nats.nix:27]: Alright. Does this config file have includes? Can you specify multiple includes? Any ability at all to keep this out of the nix store?

If I understand correctly this isn't a big deal so these things I mention are "nice to haves".

Comment (c0deaddict) [reply nixos/tests/nats.nix:27]: Apparently yes, it does support [includes](https://docs.nats.io/nats-server/configuration#include-directive). Unfortunately:

> Includes must use relative paths, and are relative to the main configuration (the one specified via the -c option):

Bit of a strange restriction, that is a deal breaker for including auth from outside of the nix store.

File: nixos/modules/services/networking/nats.nix
Comment (c0deaddict) [reply nixos/modules/services/networking/nats.nix:167]: Cool, that works!

File: nixos/modules/services/networking/nats.nix
Comment (c0deaddict) [reply nixos/modules/services/networking/nats.nix:36]: I've implemented the `StateDirectory` but am wondering what the preferred approach is now: have a user/group created like currently done, or have `DynamicUser = true` and restrict the sysadmin to always put the `dataDir` in `/var/lib`? I'm nudging towards the former because that provides more flexilibity.

File: nixos/modules/services/networking/nats.nix
Comment (SuperSandro2000) [reply nixos/modules/services/networking/nats.nix:94]: Don't we need to specially format it for the docs?

Comment (aanderse) [reply nixos/modules/services/networking/nats.nix:94]: Yes, that is what I meant... I love clickable links. Can you please make it clickable... Sorry, will be more clear next time 😄

File: nixos/modules/services/networking/nats.nix
Comment (aanderse) [reply nixos/modules/services/networking/nats.nix:31]: Dealers choice I guess. We promote `DynamicUser` but sometimes that just doesn't work.

It would be nice if we had a good design pattern in options to specify dynamic use vs NixOS provided system user vs sysadmin provided user. Currently our pattern only adequately covers either the first option **or** the last two options.

File: nixos/modules/services/networking/nats.nix
Comment (c0deaddict) [reply nixos/modules/services/networking/nats.nix]: I understood, fixed it in the new commit :smile:

File: nixos/tests/nats.nix
Comment (aanderse) [reply nixos/tests/nats.nix:27]: Understood. If you don't have any ideas I suggest leaving as you have it now.

:+1:

File: nixos/modules/services/networking/nats.nix
Comment (aanderse) [inline nixos/modules/services/networking/nats.nix:45]: Missing trailing period **or** `mkEnableOption "JetStream";`.

File: nixos/modules/services/networking/nats.nix
Comment (c0deaddict) [reply nixos/modules/services/networking/nats.nix]: Good idea, will do this! But first have to read up a bit on what these hardening options actually do.

File: nixos/modules/services/networking/nats.nix
Comment (c0deaddict) [reply nixos/modules/services/networking/nats.nix:110]: That makes sense, i've removed all hardening options for now. We can update the options in NixOS if the upstream repo agrees on the hardening changes.

Comment (aanderse) [reply nixos/modules/services/networking/nats.nix:106]: Are you familiar with `systemd.packages`? If a package includes `systemd` units you can directly use them. Not sure if it's worth it in this case, though. Just wanted to mention.

File: nixos/modules/services/networking/nats.nix
Comment (c0deaddict) [reply nixos/modules/services/networking/nats.nix]: Small update: made a PR with hardening options on the upstream repo. Got positive feedback on it. I'll wait until it's accepted and merged, then I will update this PR.

File: nixos/modules/services/networking/nats.nix
Comment (c0deaddict) [reply nixos/modules/services/networking/nats.nix:106]: TIL! Thanks for mentioning :+1:

File: nixos/modules/services/networking/nats.nix
Comment (aanderse) [reply nixos/modules/services/networking/nats.nix]: Outstanding!

File: nixos/modules/services/networking/nats.nix
Comment (c0deaddict) [reply nixos/modules/services/networking/nats.nix:31]: For simplicity sake, I would prefer `DynamicUser`, however I found another use case for the old-style users: mounting secrets with https://github.com/Mic92/sops-nix In the current implementation of this service that is not yet possible, but it can be extended that way (by having the `configFile` at a fixed location, say `/etc/nats/nats-server.conf` and then symlinking the secret part of the config to the `sops-nix` secret location).

With that in mind I would like to keep the current implementation. If there is a good design pattern to have the sysadmin choose `DynamicUser` over old-style users that would be even better.

File: nixos/modules/services/networking/nats.nix
Comment (c0deaddict) [reply nixos/modules/services/networking/nats.nix]: Unfortunately the upstream PR is not yet merged. The Nats developers have not responded for over a month. They were positive about the hardening options, and gave some feedback about *adding* some.

I also have been running a Nats server with this PR patched on nixpkgs stable on my server for over a month and haven't encountered any issues. So I think it's safe to move this PR forward and try to get it merged.

**Review by aanderse — APPROVED** [review]
Sorry, I meant to comment on this. Without being a subject matter expert LGTM :+1:

@SuperSandro2000 is there anything else you would like to see before this is merged?

**Pull request merged**
